File: integrations/python/posthog/src/axora_logger_posthog/posthog_sink.py
from __future__ import annotations

from typing import Any, Protocol

from axora_logger import Sink


class PosthogClientLike(Protocol):
    def capture(self, distinct_id: str, event: str, properties: dict[str, Any], timestamp: str | None = None) -> None: ...

    def shutdown(self) -> None: ...


class PosthogSink(Sink):
    def __init__(self, client: PosthogClientLike) -> None:
        self._client = client

    async def export(self, event: dict[str, Any]) -> None:
        distinct_id = event["context"].get("user_id") or f"service:{event['service']}"
        self._client.capture(
            distinct_id,
            event["operation"],
            {
                **event["context"],
                "axora_event_id": event["event_id"],
                "axora_service": event["service"],
                "status": event["status"],
                "level": event["level"],
                "duration_ms": event["duration_ms"],
                **({"error_message": event["error"]["message"]} if event["error"]["message"] else {}),
            },
            timestamp=event["timestamp_start"],
        )

    async def flush(self) -> None:
        self._client.shutdown()

File: integrations/java/posthog/src/main/java/com/axora/logger/posthog/PosthogSink.java
package com.axora.logger.posthog;

import com.axora.logger.WideEventPayload;
import com.axora.logger.sinks.Sink;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public final class PosthogSink implements Sink {
  private final PosthogClientLike client;

  public PosthogSink(PosthogClientLike client) {
    this.client = client;
  }

  @Override
  public CompletableFuture<Void> export(WideEventPayload event) {
    String distinctId = event.context().containsKey("user_id")
        ? String.valueOf(event.context().get("user_id"))
        : "service:" + event.service();

    Map<String, Object> properties = new LinkedHashMap<>(event.context());
    properties.put("axora_event_id", event.eventId());
    properties.put("axora_service", event.service());
    properties.put("status", event.status());
    properties.put("level", event.level());
    properties.put("duration_ms", event.durationMs());
    if (event.error().message() != null) {
      properties.put("error_message", event.error().message());
    }

    client.capture(distinctId, event.operation(), properties, Instant.parse(event.timestampStart()));
    return CompletableFuture.completedFuture(null);
  }

  @Override
  public CompletableFuture<Void> flush() {
    client.shutdown();
    return CompletableFuture.completedFuture(null);
  }
}
